### ISSUE TITLE
Edit and see edited notes Positive Interoperability Tracker

### ISSUE THREAD

**Issue by alltheseas** [nostrability/nostrability#118]
**app**|**See edited notes**|**Edit Posts**|**comment**
-----|-----|----- |-----
damus|❌|?|
amethyst|✅|✅|

**Comment by alltheseas** [comment]
@vitorpamplona what is the spec for edited notes?

**Comment by vitorpamplona** [comment]
PR: https://github.com/nostr-protocol/nips/pull/1090

**Comment by alltheseas** [comment]
Is the above spec constrained to kind1 only?

**Comment by vitorpamplona** [comment]
Yep, only the `.content` of kind 1 notes.

**Comment by mikedilger** [comment]
There was a competing edit spec by Staab https://github.com/nostr-protocol/nips/pull/1091 which gossip implements. Everybody can see the annotations because they are just replies with a special marking, and for clients that implement it they see it inside the original note.

Maybe we can change the table here to specify which clients implement which spec, instead of one issue per different spec, since these kind of compete?

**Comment by vitorpamplona** [comment]
IMO Annotations and edits are different things, they fix different issues. They are not a replacement for one another. Both NIPs should be implemented and offered to users. 

Annotations were offered as a replacement for edits because hodlbod doesn't like edits from an academic/philosophical point of view. Which is a fine position for a Client, but bad for a protocol.

**Comment by mikedilger** [comment]
I have no problem with edits from an academic/philosophical point of view.  In fact I've long felt that all events should be referenced by a pubkey and d-tag, and not the hash of their contents, so that they could all be replaced at the same address and there would be no replacement logic necessary.  And yes I know your edits are not replacements, they are additions.  But real edits would be replacements.  And annotations are real additions.

**Comment by fiatjaf** [comment]
If most people start using this it becomes mandatory, and then it raises the implementation difficulty bar for `kind:1` clients by a lot. And if you want to display the edited events in a feed that would require an absurd amount of relay queries and would slow everything down.

I feel like Nostr should strive to keep the `kind:1` bar very simple and easy to implement and let complicated stuff be done in other kinds. `kind:1` is the most crucial place to maintain Nostr decentralization, and having lots of clients with different approaches is crucial for that. This kind of change will necessarily limit the number of clients and the number of approaches clients can have to fetching and displaying data, and it may also hurt performance and therefore UX.

The annotations spec is less bad because annotations are not expected, for example, to be displayed in a long running live or historical feed of notes.

**Comment by vitorpamplona** [comment]
> If most people start using this it becomes mandatory, and then it raises the implementation difficulty bar for kind:1 clients by a lot. 

I don't think this is mandatory for kind 1 clients. We have been running edits on Amethyst for over 9 months now and Amethyst users don't give a shit if the other client doesn't implement edits. They edit anyway. Non-amethyst users are aware that they do not see the edits. And they also seem fine with it. Edits is just a feature like replies. You can do a kind 1 without showing any replies and it will be ok. 

> The annotations spec is less bad because annotations are not expected, for example, to be displayed in a long running live or historical feed of notes.

The difference is insignificant. If you get many edits, you would have gotten many annotation events as well. But again, annotations are not a replacement for edits. They don't fix your text in the way users want. They just offer additional points.

**Comment by alltheseas** [comment]
On Amethyst you can cycle between original post, and edited versions by tapping on the post. 

![Frame 27](https://github.com/user-attachments/assets/01e6c902-5d91-4fb8-9e98-5a7c36ccb716)

Whereas annotations are displayed altogether. See https://github.com/nostrability/nostrability/issues/120#issuecomment-2442526222

**Comment by jb55** [comment]
> In fact I've long felt that all events should be referenced by a pubkey and d-tag, and not the hash of their contents, so that they could all be replaced at the same address and there would be no replacement logic necessary

this is a really bad idea, you would never know what people actually replied to.

**Comment by alltheseas** [comment]
It seems counterintuitive to me that edit post, and annotations are functionally identical from the end user's perspective, yet they are not interoperable. 

Both seem to retain the original note, and subsequent edits, annotations (see screenshots https://github.com/nostrability/nostrability/issues/118#issuecomment-2442544150 and https://github.com/nostrability/nostrability/issues/120#issuecomment-2442526222)

Is there no way to reconcile the two specs?

**Comment by alltheseas** [comment]
Asked nostr about UI

https://njump.me/nevent1qvzqqqqqqypzq96n3hp2vfmf6z2y8uvvxl97xk86kkalnqghx4p25lzl79c76a7yqyfhwumn8ghj7am0wsh82arcduhx7mn99uq3xamnwvaz7tmsw4e8qmr9wpskwtn9wvhsqg8zc2rhpens477zgzpqy4yzsa6t4wleu6ggu8s4qatf4wqy3ju00sylu22p

**Comment by alltheseas** [comment]
> I don't think this is mandatory for kind 1 clients

Agreed, and I did not intend to make impression or suggestion of this being mandatory.

**Comment by alltheseas** [comment]
Bluesky remains incompatible with edits

![image](https://github.com/user-attachments/assets/9aacb788-7385-4160-8ae0-0f744fc16655)

**Comment by alltheseas** [comment]
@fiatjaf has a new edit notes proposal out: https://github.com/nostr-protocol/nips/pull/1565 

Consider holding off on implementing edits until consensus is reached

**Comment by alltheseas** [comment]
https://njump.me/note1924pjx5c60ej0ujvd6r78rn3xrl99ltkdf66l9tplanygfk7flzstqe3q7

Consider this unhappy path

> On Facebook, I see edit scams all the time. Someone makes a post about a missing dog or something, tons of people share it around, and later it gets edited to something trying to sell $10 sunglasses. As ridiculous a technique it may be, it hinges on people's emotions to work. Whether they sell any sunglasses or not is beside the point. Over time, the people sharing those posts become reluctant of what may be legitimate information & the actual missing dog gets no coverage.

**Comment by alltheseas** [comment]
how github displays edits visually 
![Screenshot from 2024-11-26 10-22-49](https://github.com/user-attachments/assets/0f3c2315-eda8-4eb6-8c51-c84214c887f7)

**Comment by fiatjaf** [comment]
https://fiatjaf.com/ad84e3b3.html

**Comment by vitorpamplona** [comment]
One of the most interesting learnings from this edit conversation is knowing how much devs want to control what users can and cannot do in the network. So much that they spend days trying to break implementations just to control users.

**Comment by alltheseas** [comment]
Here is an example proofread and 10 second undo user story: https://github.com/damus-io/damus/issues/801

**Comment by staab** [comment]
How Coracle handles undo via delayed send:

https://github.com/user-attachments/assets/a1c76905-899c-4827-9ef0-d6facba0e7c5

Flotilla:

https://github.com/user-attachments/assets/62375773-abaa-43fa-baab-51d8b102ee15

**Comment by alltheseas** [comment]
Amethyst sees edited note gibberish by @fiatjaf 

https://njump.me/nevent1qqsqcc7c4r32pl577hs3al000p2shtf58fjvvs7hsqeudp82s2akn4qprpmhxue69uhhyetvv9ujuumwdae8gtnnda3kjctvqy2hwumn8ghj7un9d3shjtnddaehgu3wwp6kyqg4waehxw309ahx7um5wghx66tvda6jumr0dsq3uamnwvaz7tmwdaehgu3dwfjkccte9ejx2un9ddex7umn9ekk2ww9hwd

![image](https://github.com/user-attachments/assets/403f144f-6146-428c-a818-a602254c0be8)

**Comment by hzrd149** [comment]
@alltheseas noStrudel shows the edits when you viewing a specific note but it does not proactively fetch them or show them in the main UI

I tend to agree with @fiatjaf that edits are not good. so I've implemented them as "corrections" that only show when a user is viewing a note
I would like to open it up in the future to allow other users to correct your notes but thats not on the todo list at the moment
